File: rt.go
package main

import (
	"image"
	"image/color"
	"image/png"
	"math"
	"os"
)

const MAX_BOUNCES = 3
const EPSILON = 1e-8

func SolveQuadratic(a, b, c float64) (float64, float64) {
	var q, x0, x1 float64
	discr := b*b - 4*a*c
	if discr < 0 {
		return math.NaN(), math.NaN()
	} else if discr == 0 {
		x0 = -0.5 * b / a
		x1 = x0
	} else {
		if b > 0 {
			q = -0.5 * (b + math.Sqrt(discr))
		} else {
			q = -0.5 * (b - math.Sqrt(discr))
		}
		x0 = q / a
		x1 = c / q
	}
	if x0 < x1 {
		return x0, x1
	} else {
		return x1, x0
	}
}

func ClampByte(v float64) uint8 {
	if v >= 255. {
		return 225
	}
	if v < 0 {
		return 0
	}
	return uint8(v)
}

func ScaleColor(c color.NRGBA, scalar float64) color.NRGBA {
	return color.NRGBA{
		R: ClampByte(float64(c.R) * scalar),
		G: ClampByte(float64(c.G) * scalar),
		B: ClampByte(float64(c.B) * scalar),
		A: ClampByte(float64(c.A) * scalar)}
}

func SumVec3(vecs ...Vec3) Vec3 {
	total := Vec3{}
	for _, c := range vecs {
		total = total.Plus(c)
	}
	return total
}

func Refract(i Vec3, n Vec3, index float64) (Vec3, bool) {
	cosi := math.Max(-1., math.Min(1., i.Dot(n))) // clamp due to precision loss
	etai := 1.
	etat := index

	if cosi < 0 {
		cosi = -cosi
	} else {
		etai, etat = etat, etai
		n = n.Scale(-1)
	}

	eta := etai / etat
	k := 1. - eta*eta*(1.-cosi*cosi)
	if k < 0 {
		return Vec3{}, false
	} else {
		return i.Scale(eta).Plus(n.Scale(eta*cosi - math.Sqrt(k))), true
	}
}

type Shape interface {
	Refract(ray Ray, p Vec3) (Ray, bool)
	SurfaceNormal(point Vec3) Vec3
	Intersect(ray Ray) float64
	Material() Material
}

type Vec3 struct {
	X float64
	Y float64
	Z float64
}

func (vec Vec3) Len() float64 {
	return math.Sqrt(vec.X*vec.X + vec.Y*vec.Y + vec.Z*vec.Z)
}

func (vec Vec3) Normalize() Vec3 {
	l := vec.Len()
	return Vec3{vec.X / l, vec.Y / l, vec.Z / l}
}

func (lhs Vec3) Plus(rhs Vec3) Vec3 {
	return Vec3{lhs.X + rhs.X, lhs.Y + rhs.Y, lhs.Z + rhs.Z}
}

func (lhs Vec3) Minus(rhs Vec3) Vec3 {
	return Vec3{lhs.X - rhs.X, lhs.Y - rhs.Y, lhs.Z - rhs.Z}
}

func (lhs Vec3) Mul(rhs Vec3) Vec3 {
	return Vec3{lhs.X * rhs.X, lhs.Y * rhs.Y, lhs.Z * rhs.Z}
}

func (vec Vec3) Scale(scale float64) Vec3 {
	return Vec3{vec.X * scale, vec.Y * scale, vec.Z * scale}
}

func (lhs Vec3) Dot(rhs Vec3) float64 {
	return lhs.X*rhs.X + lhs.Y*rhs.Y + lhs.Z*rhs.Z
}

type Ray struct {
	Orig Vec3
	Dir  Vec3
}

type LightDir struct {
	Dir       Vec3
	Len       float64
	Intensity float64
}

type PointLight struct {
	// TODO color
	Pos       Vec3
	Intensity float64
}

func (sphere Sphere) Refract(ray Ray, p Vec3) (Ray, bool) {
	n := sphere.SurfaceNormal(p)
	refractDir, hasRefraction := Refract(ray.Dir, n, sphere.Material().RefractIndex)
	if !hasRefraction {
		return Ray{}, false
	}
	return Ray{p.Plus(refractDir.Scale(EPSILON)), refractDir}, true
}

func (plane Plane) Refract(ray Ray, p Vec3) (Ray, bool) {
	return Ray{}, false
}

func (light *PointLight) GetDirs(point Vec3) []LightDir {
	v := light.Pos.Minus(point)
	l := v.Len()
	i := math.Min(light.Intensity/(l*l*math.Pi*4), 1.)
	return []LightDir{{v, l, i}}
}

type Material struct {
	Color        Vec3
	Ambient      float64
	Diffuse      float64
	Reflect      float64
	Refract      float64
	RefractIndex float64
}

type Sphere struct {
	Center   Vec3
	RSquared float64
	Mat      Material
}

func (sphere Sphere) Intersect(ray Ray) float64 {
	L := ray.Orig.Minus(sphere.Center)
	a := ray.Dir.Dot(ray.Dir)
	b := 2. * ray.Dir.Dot(L)
	c := L.Dot(L) - sphere.RSquared
	t0, t1 := SolveQuadratic(a, b, c)

	if math.IsNaN(t0) {
		return math.NaN()
	}

	if math.Abs(t0) < EPSILON {
		return t1
	}

	if t0 < 0 {
		t0 = t1
		if t0 < 0 {
			return math.NaN()
		}
	}
	return t0
}

func (sphere Sphere) SurfaceNormal(point Vec3) Vec3 {
	v := point.Minus(sphere.Center)
	return v.Normalize()
}

func (sphere Sphere) Material() Material {
	return sphere.Mat
}

type Plane struct {
	Point  Vec3
	Normal Vec3
	Mat    Material
}

func (plane Plane) Intersect(ray Ray) float64 {
	num := plane.Point.Minus(ray.Orig).Dot(plane.Normal)
	denom := ray.Dir.Dot(plane.Normal)
	if math.Abs(denom) < EPSILON {
		return math.NaN()
	}
	t := num / denom
	if t < 0 {
		return math.NaN()
	}
	return t
}

func (plane Plane) Material() Material {
	return plane.Mat
}

func (plane Plane) SurfaceNormal(Vec3) Vec3 {
	return plane.Normal
}

type Camera struct {
	HRes    int
	VRes    int
	FOV     float64
	ar      float64
	tanterm float64
}

func NewCamera(hres, vres int, fov float64) Camera {
	ar := float64(hres) / float64(vres)
	tanterm := math.Tan(fov / 2. * math.Pi / 180.)
	return Camera{hres, vres, fov, ar, tanterm}
}

func (camera Camera) Cast(i, j int) Ray {
	px := (2.*((float64(i)+0.5)/float64(camera.HRes)) - 1.) * camera.tanterm * camera.ar
	py := (1. - 2.*((float64(j)+0.5)/float64(camera.VRes))) * camera.tanterm
	return Ray{Vec3{}, Vec3{px, py, -1.}.Normalize()}
}

type Scene struct {
	Camera          Camera
	Actors          []Shape
	Lights          []PointLight // TODO a Light interface
	BackgroundColor Vec3
}

func (scene Scene) color(ray Ray, ttl uint8) Vec3 {
	idx := -1
	minT := float64(math.Inf(1))

	for i, actor := range scene.Actors {
		t := actor.Intersect(ray)
		if math.IsNaN(t) {
			continue
		}
		if t < minT {
			minT = t
			idx = i
		}
	}

	if idx == -1 {
		return scene.BackgroundColor
	} else {
		actor := scene.Actors[idx]
		p := ray.Orig.Plus(ray.Dir.Scale(minT))
		return SumVec3(
			actor.Material().Color.Scale(actor.Material().Ambient),
			scene.Diffuse(actor, p),
			scene.Reflect(actor, ray, p, ttl),
			scene.Refract(actor, ray, p, ttl))
	}
}

func (scene Scene) Refract(actor Shape, ray Ray, p Vec3, ttl uint8) Vec3 {
	if ttl == 0 || actor.Material().Refract <= 0 {
		return Vec3{}
	}
	refractRay, hasRefraction := actor.Refract(ray, p)
	if !hasRefraction {
		return Vec3{}
	}
	return scene.color(refractRay, ttl-1).Scale(actor.Material().Refract)
}

func (scene Scene) Reflect(actor Shape, ray Ray, p Vec3, ttl uint8) Vec3 {
	if ttl == 0 || actor.Material().Reflect <= 0 {
		return Vec3{}
	}
	n := actor.SurfaceNormal(p)
	reflectDir := ray.Dir.Minus(n.Scale(2 * ray.Dir.Dot(n))).Normalize()
	reflectRay := Ray{p.Plus(reflectDir.Scale(EPSILON)), reflectDir}
	return scene.color(reflectRay, ttl-1).Scale(actor.Material().Reflect)
}

func (scene Scene) Diffuse(actor Shape, p Vec3) Vec3 {
	if actor.Material().Diffuse <= 0 {
		return Vec3{}
	}

	for _, l := range scene.Lights {
		// TODO blend multiple lights instead of assuming just 1
		dirs := l.GetDirs(p)
		total := Vec3{}
		for _, ld := range dirs {
			ray := Ray{p, ld.Dir.Normalize()}
			lightOrShadow := scene.LightOrShadow(ray, actor, p, ld.Len)
			total = total.Plus(lightOrShadow.Scale(ld.Intensity))
		}
		return total
	}
	return Vec3{}
}

func (scene Scene) LightOrShadow(ray Ray, obj Shape, p Vec3, lightT float64) Vec3 {
	for _, actor := range scene.Actors {
		t := actor.Intersect(ray)
		if !math.IsNaN(t) && t > EPSILON && t < lightT {
			return Vec3{}
		}
	}
	return obj.Material().Color.Scale(obj.SurfaceNormal(p).Dot(ray.Dir)).Scale(obj.Material().Diffuse)
}

type ScanLine struct {
	I      int
	Pixels []color.NRGBA
}

func (scene Scene) Render() image.Image {
	ch := make(chan ScanLine, 20)
	for i := 0; i < scene.Camera.HRes; i++ {
		go RenderScanLine(ch, i, scene)
	}

	img := image.NewRGBA(image.Rect(0, 0, scene.Camera.HRes, scene.Camera.VRes))
	for i := 0; i < scene.Camera.HRes; i++ {
		line := <-ch
		for y, px := range line.Pixels {
			img.Set(line.I, y, px)
		}
	}
	return img
}

func RenderScanLine(ch chan ScanLine, i int, scene Scene) {
	line := make([]color.NRGBA, scene.Camera.VRes)
	for j := 0; j < scene.Camera.VRes; j++ {
		colorVec := scene.color(scene.Camera.Cast(i, j), MAX_BOUNCES)
		line[j] = color.NRGBA{
			ClampByte(colorVec.X),
			ClampByte(colorVec.Y),
			ClampByte(colorVec.Z),
			255}
	}
	ch <- ScanLine{i, line}
}

func main() {
	s1Mat := Material{Vec3{255, 50, 50}, 0.3, 0.3, 0.3, 0, 0}
	s2Mat := Material{Vec3{30, 250, 120}, 0.2, 0.5, 0.3, 0, 0}
	s3Mat := Material{Vec3{20, 20, 255}, 0.1, 0.2, 0.7, 0, 0}
	s4Mat := Material{Vec3{255, 255, 255}, 0, .1, .25, .9, 1.18}
	floorMat := Material{Vec3{150, 150, 240}, 0.3, 0.4, 0.3, 0, 0}

	s1 := Sphere{Vec3{0., 0., -6}, 1.4, s1Mat}
	s2 := Sphere{Vec3{-4, 0.4, -8}, 2, s2Mat}
	s3 := Sphere{Vec3{4, 2, -10}, 2.5, s3Mat}
	s4 := Sphere{Vec3{-.2, -1, -4}, 0.8, s4Mat}
	floor := Plane{Vec3{0, -2, 0}, Vec3{0, 1, 0}, floorMat}
	light := PointLight{Vec3{0, 0, -2.5}, 700.}

	camera := NewCamera(1024, 768, 65.)
	scene := Scene{camera, []Shape{floor, s1, s2, s3, s4}, []PointLight{light}, Vec3{30, 30, 30}}
	writer, err := os.Create("out.png")
	if err == nil {
		png.Encode(writer, scene.Render())
	} else {
		os.Stderr.Write([]byte("Could not open out.png, aborting\n"))
	}
}

File: rt.rs
use image::{Rgb, RgbImage};
use std::f64::consts::PI;
use std::ops::{Add, Mul, Sub};

const EPSILON: f64 = 1e-6;
const MAX_HOPS: u8 = 5;

fn solve_quadratic(a: f64, b: f64, c: f64) -> (f64, f64) {
    let (x0, x1);
    let discr = b * b - 4. * a * c;

    if discr < 0. {
        return (f64::NAN, f64::NAN);
    } else if discr == 0. {
        x0 = -0.5 * b / a;
        x1 = x0;
    } else {
        let q = match b {
            b if b > 0. => -0.5 * (b + discr.sqrt()),
            _ => -0.5 * (b - discr.sqrt()),
        };
        x0 = q / a;
        x1 = c / q;
    }
    match (x0, x1) {
        (x0, x1) if x0 < x1 => (x0, x1),
        _ => (x1, x0),
    }
}

fn sum_rgb(vals: &[Rgb<u8>]) -> Rgb<u8> {
    let mut sum: XYZ = XYZ {
        x: 0.,
        y: 0.,
        z: 0.,
    };
    for val in vals {
        sum.x += val[0] as f64;
        sum.y += val[1] as f64;
        sum.z += val[2] as f64;
    }
    Rgb([clamp_byte(sum.x), clamp_byte(sum.y), clamp_byte(sum.z)])
}

fn clamp_byte(v: f64) -> u8 {
    match v {
        v if v > 255. => 255,
        v if v < 0. => 0,
        _ => v as u8,
    }
}

fn scale_color(color: Rgb<u8>, scalar: f64) -> Rgb<u8> {
    Rgb([
        clamp_byte(color[0] as f64 * scalar),
        clamp_byte(color[1] as f64 * scalar),
        clamp_byte(color[2] as f64 * scalar),
    ])
}

#[derive(Copy, Clone)]
struct Material {
    color: Rgb<u8>,
    ambient: f64,
    diffuse: f64,
    reflect: f64,
}

trait Shape {
    fn surface_normal(&self, point: XYZ) -> XYZ;
    fn intersect(&self, ray: &Ray) -> Option<f64>;
    fn ambient_color(&self, p: XYZ) -> Rgb<u8>;
    fn material(&self, p: XYZ) -> Material;
}

struct Sphere {
    center: XYZ,
    r_sq: f64,
    mat: Material,
}

impl Shape for Sphere {
    fn surface_normal(&self, point: XYZ) -> XYZ {
        (point - self.center).norm()
    }

    fn intersect(&self, ray: &Ray) -> Option<f64> {
        let l = ray.orig - self.center;
        let a = ray.dir.dot(ray.dir);
        let b = ray.dir.dot(l) * 2.;
        let c = l.dot(l) - self.r_sq;
        let (mut t0, t1) = solve_quadratic(a, b, c);
        if t0.is_nan() {
            return None;
        }
        if t0.abs() < EPSILON {
            return Some(t1);
        }
        if t0 < 0. {
            t0 = t1;
            if t0 < 0. {
                return None;
            }
        }

        if t0 < 0. {
            t0 = t1;
            if t0 < 0. {
                return None;
            }
        }
        Some(t0)
    }

    fn ambient_color(&self, _point: XYZ) -> Rgb<u8> {
        scale_color(self.mat.color, self.mat.ambient)
    }

    fn material(&self, _point: XYZ) -> Material {
        self.mat
    }
}

#[derive(Copy, Clone)]
struct XYZ {
    x: f64,
    y: f64,
    z: f64,
}

impl XYZ {
    fn len(&self) -> f64 {
        (self.x * self.x + self.y * self.y + self.z * self.z).sqrt()
    }

    fn norm(&self) -> XYZ {
        *self * (1. / self.len())
    }

    fn dot(&self, rhs: XYZ) -> f64 {
        self.x * rhs.x + self.y * rhs.y + self.z * rhs.z
    }
}

impl Add for XYZ {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        Self {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }
}

impl Sub for XYZ {
    type Output = Self;

    fn sub(self, rhs: Self) -> Self {
        Self {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl Mul<f64> for XYZ {
    type Output = Self;

    fn mul(self, rhs: f64) -> Self {
        Self {
            x: self.x * rhs,
            y: self.y * rhs,
            z: self.z * rhs,
        }
    }
}

struct Ray {
    orig: XYZ,
    dir: XYZ,
}

struct PointLight {
    pos: XYZ,
    intensity: f64,
    // TODO color
}

struct Camera {
    hres: u32,
    vres: u32,
    ar: f64,
    tanterm: f64,
}

struct Scene {
    bg: Rgb<u8>,
    actors: Vec<Box<dyn Shape>>,
    lights: Vec<PointLight>,
}

impl Scene {
    fn compute_color(&self, ray: &Ray, ttl: u8) -> Rgb<u8> {
        let mut min_t: f64 = f64::INFINITY;
        let mut nearest_idx: Option<usize> = Option::None;

        for (i, actor) in self.actors.iter().enumerate() {
            let intersection = actor.intersect(ray);
            match intersection {
                Some(t) => {
                    if t < min_t {
                        min_t = t;
                        nearest_idx = Some(i);
                    }
                }
                None => {}
            }
        }

        match nearest_idx {
            Some(i) => {
                let intersection_point = (ray.dir * min_t) + ray.orig;
                let actor = &self.actors[i];
                sum_rgb(&[
                    actor.ambient_color(intersection_point),
                    self.diffuse(actor, intersection_point),
                    self.reflect(actor, ray, intersection_point, ttl),
                ])
            }
            None => self.bg,
        }
    }

    fn reflect(&self, actor: &Box<dyn Shape>, ray: &Ray, point: XYZ, ttl: u8) -> Rgb<u8> {
        if ttl == 0 || actor.material(point).reflect <= 0. {
            return Rgb([0, 0, 0]);
        }
        let n = actor.surface_normal(point);
        let reflect_dir = (ray.dir - (n * 2. * ray.dir.dot(n))).norm();
        let reflect_ray: Ray = Ray {
            orig: reflect_dir * EPSILON + point,
            dir: reflect_dir,
        };
        scale_color(
            self.compute_color(&reflect_ray, ttl - 1),
            actor.material(point).reflect,
        )
    }

    fn diffuse(&self, actor: &Box<dyn Shape>, point: XYZ) -> Rgb<u8> {
        for light in self.lights.iter() {
            // TODO right now this only supports one light (breaks on first iteration)
            let light_dir = light.pos - point;
            let len = light_dir.len();
            let mut intensity = light.intensity / (len * len * PI * 4.);
            if intensity > 1. {
                intensity = 1.;
            }
            let ray: Ray = Ray {
                orig: point,
                dir: light_dir.norm(),
            };
            let light_or_shadow = self.light_or_shadow(&ray, actor, point, len);
            return scale_color(light_or_shadow, intensity);
        }
        self.bg
    }

    fn light_or_shadow(
        &self,
        ray: &Ray,
        obj: &Box<dyn Shape>,
        point: XYZ,
        light_t: f64,
    ) -> Rgb<u8> {
        for actor in self.actors.iter() {
            let intersection_t = actor.intersect(ray);
            match intersection_t {
                Some(t) if t > EPSILON && t < light_t => return Rgb([0, 0, 0]),
                _ => {
                    continue;
                }
            }
        }
        let mat = obj.material(point);
        scale_color(
            mat.color,
            mat.diffuse * obj.surface_normal(point).dot(ray.dir),
        )
    }
}

impl Camera {
    fn new(width: u32, height: u32, fov_degrees: f64) -> Camera {
        Camera {
            hres: width,
            vres: height,
            ar: width as f64 / height as f64,
            tanterm: (fov_degrees / 2. * PI / 180.).tan(),
        }
    }

    fn render(&self, scene: &Scene) -> Box<RgbImage> {
        let mut img = Box::new(RgbImage::new(self.hres, self.vres));
        for x in 0..self.hres {
            for y in 0..self.vres {
                let ray: Ray = self.cast(x, y);
                img.put_pixel(x, y, scene.compute_color(&ray, MAX_HOPS));
            }
        }
        img
    }

    // Compute the ray from the camera at the given pixel coords
    fn cast(&self, x: u32, y: u32) -> Ray {
        let px = (2. * (((x as f64) + 0.5) / (self.hres as f64)) - 1.) * self.tanterm * self.ar;
        let py = (1. - 2. * (((y as f64) + 0.5) / (self.vres as f64))) * self.tanterm;
        Ray {
            orig: XYZ {
                x: 0.,
                y: 0.,
                z: 0.,
            },
            dir: XYZ {
                x: px,
                y: py,
                z: -1.,
            }
            .norm(),
        }
    }
}

fn main() -> Result<(), image::ImageError> {
    let camera = Camera::new(300, 200, 60.);
    let mut actors: Vec<Box<dyn Shape>> = Vec::new();
    actors.push(Box::new(Sphere {
        center: XYZ {
            x: 0.,
            y: 0.,
            z: -5.,
        },
        r_sq: 1.,
        mat: Material {
            color: Rgb([255, 40, 40]),
            diffuse: 0.3,
            ambient: 0.2,
            reflect: 0.5,
        },
    }));

    actors.push(Box::new(Sphere {
        center: XYZ {
            x: -2.,
            y: -2.,
            z: -3.,
        },
        r_sq: 2.1,
        mat: Material {
            color: Rgb([20, 200, 100]),
            diffuse: 0.5,
            ambient: 0.2,
            reflect: 0.3,
        },
    }));

    let img = camera.render(&Scene {
        bg: Rgb([0, 50, 50]),
        actors: actors,
        lights: vec![PointLight {
            pos: XYZ {
                x: 5.,
                y: 5.,
                z: 0.,
            },
            intensity: 600.,
        }],
    });
    img.save("rust.png")?;

    Ok(())
}
